src/codecs/bmp/decoder.rs

```diff
@@ -1579,53 +1437,4 @@ mod test {
         let mut buf: Vec<u8> = vec![0; 8 * 8 * 3];
         decoder.read_rect(0, 0, 8, 8, &mut *buf).unwrap();
     }
-
-    #[test]
-    fn read_rle_too_short() {
-        let data = vec![
-            0x42, 0x4d, 0x04, 0xee, 0xfe, 0xff, 0xff, 0x10, 0xff, 0x00, 0x04, 0x00, 0x00, 0x00,
-            0x7c, 0x00, 0x00, 0x00, 0x0c, 0x41, 0x00, 0x00, 0x07, 0x10, 0x00, 0x00, 0x01, 0x00,
-            0x04, 0x00, 0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x0d, 0x00, 0x00, 0x00,
-            0x00, 0x80, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xff, 0xff, 0xff, 0xff, 0xfe, 0x21,
-            0xff, 0x00, 0x66, 0x61, 0x72, 0x62, 0x66, 0x65, 0x6c, 0x64, 0x00, 0x00, 0x00, 0x00,
-            0x00, 0x00, 0x00, 0xff, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
-            0xff, 0xd8, 0xff, 0x00, 0x00, 0x19, 0x51, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
-            0x00, 0x00, 0x00, 0x10, 0x00, 0x00, 0x00, 0x00, 0x00, 0xfa, 0xff, 0x00, 0x00, 0x00,
-            0x00, 0x01, 0x00, 0x11, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x0f, 0x00,
-            0x00, 0x00, 0x00, 0x2d, 0x31, 0x31, 0x35, 0x36, 0x00, 0xff, 0x00, 0x00, 0x52, 0x3a,
-            0x37, 0x30, 0x7e, 0x71, 0x63, 0x91, 0x5a, 0x04, 0x00, 0x10, 0x00, 0x00, 0x00, 0x00,
-            0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00,
-            0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x2d, 0x35, 0x37, 0x00, 0xff, 0x00, 0x00, 0x52,
-            0x3a, 0x37, 0x30, 0x7e, 0x71, 0x63, 0x91, 0x5a, 0x04, 0x05, 0x3c, 0x00, 0x00, 0x11,
-            0x00, 0x5d, 0x7a, 0x82, 0xb7, 0xca, 0x2d, 0x31, 0xff, 0xff, 0xc7, 0x95, 0x33, 0x2e,
-            0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x7c, 0x00,
-            0x20, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x20, 0x66, 0x00, 0x4d,
-            0x4d, 0x00, 0x2a, 0x00,
-        ];
-
-        let decoder = BmpDecoder::new(Cursor::new(&data)).unwrap();
-        let mut buf = vec![0; usize::try_from(decoder.total_bytes()).unwrap()];
-        assert!(decoder.read_image(&mut buf).is_err());
-    }
-
-    #[test]
-    fn test_extend_buffer() {
-        // (input, extend_to)
-        let test_cases: &[(&[u8], usize)] = &[
-            (&[0, 1, 2, 3, 4, 5, 6, 7, 8, 9], 20),
-            (&[0, 1, 2, 3, 4, 5, 6, 7, 8, 9], 13),
-            (&[0, 1, 2, 3, 4, 5, 6, 7, 8, 9], 50),
-            (&[0, 1, 2, 3], 5),
-            (&[0, 1, 2, 3], 15),
-            (&[0, 1, 2, 3], 50),
-        ];
-
-        for test_case in test_cases {
-            let mut as_vec = test_case.0.to_vec();
-            let extended_by = test_case.1 - test_case.0.len();
-            let allocated = extend_buffer(&mut as_vec, test_case.1, false);
-            assert_eq!(allocated.len(), extended_by);
-            assert_eq!(&as_vec[extended_by..], test_case.0);
-        }
-    }
```

Comment:
`test_extend_buffer` was removed because the extend_buffer functionality is no-longer needed now that the decoder writes directly into the output buffer (i.e. it tests a now removed method).

I added back `read_rle_too_short`. This is related to my comment above. The old code was inconsistent about when it returned an error about incomplete data or silently returned a partial image. The determination was was based on some weird logic related to how the total size of the image and how much had been decoded by the time the end of image marker was hit. Instead we now always return the partial image because that seemed less likely to break existing users.